### PR TITLE
[FSDP2] Add opt-in gradient dtype independent of parameter dtype

### PR DESCRIPTION
**Pull request by xiuhu17**
Human request (verbatim): Support higher precision for fsdp wgrad.

@ppl-ai

AI-assisted implementation and description.

A custom backward can produce FP32 weight gradients, but a BF16 FSDP2 parameter normally rounds them to BF16 before reduction. `reduce_dtype=torch.float32` cannot recover that precision.

This PR adds optional `MixedPrecisionPolicy.grad_dtype` to preserve higher-precision gradients at the unsharded receiver and in the final sharded `.grad`:

```python
MixedPrecisionPolicy(
    param_dtype=torch.bfloat16,
    reduce_dtype=torch.float32,
    grad_dtype=torch.float32,
)
```

`grad_dtype=None` preserves existing behavior. If `reduce_dtype` is omitted, it defaults to the requested gradient dtype. Parameter storage and all-gather behavior are unchanged. The implementation also preserves gradient dtype metadata and values across parameter/device conversions, including partial conversion failures.

Backward must explicitly produce FP32 gradients; this does not change ordinary Linear backward GEMM precision. Optimizers must support the selected parameter/gradient dtypes. Compiler support is tracked separately in #197381.

### Application numerical results

Relative L2 ranges below compare the final reduced/accumulated gradient shard or FP32 optimizer master with an independently constructed reference, across topology, scheduling, reshard, and step combinations. The gradient reference uses the **same BF16 hidden and dLogits operands**, an explicit FP32-output GEMM, and FP32 SUM all-reduction. It is not an FP64 backward or an all-FP32 model reference. These application harnesses are not part of the upstream test suite.

| Native execution | Objective | Native wgrad rel-L2 | Adapter wgrad rel-L2 | Disabled-policy wgrad rel-L2 | Native FP32 master rel-L2 |
| --- | --- | --- | --- | --- | --- |
| A: eager | Cross-entropy | `5.801e-08`–`6.625e-08` | `5.039e-08`–`6.089e-08` | `1.861e-03`–`1.931e-03` | `1.641e-08`–`2.635e-08` |
| A: eager | Clipped policy | `5.465e-08`–`6.285e-08` | `4.714e-08`–`6.080e-08` | `1.858e-03`–`1.909e-03` | `1.675e-08`–`2.314e-08` |
| B: Inductor projection | Cross-entropy | `5.039e-08`–`6.089e-08` | `5.039e-08`–`6.089e-08` | `1.861e-03`–`1.931e-03` | `1.634e-08`–`2.632e-08` |
| B: Inductor projection | Clipped policy | `4.714e-08`–`6.080e-08` | `4.714e-08`–`6.080e-08` | `1.858e-03`–`1.909e-03` | `1.616e-08`–`2.277e-08` |

A denotes the historical eager validation revision (`400b12300d026d40c42b3ffba30dd92873f35288`); B adds the compiler follow-up in #197381, which is outside this PR. These application matrices were not rerun after the later conversion fixes.

### Validation

- Initial FSDP regression suite: **33 passed**. Each application matrix covered **16 configurations and 32 optimizer-step records**.
- Latest targeted policy/conversion suite: **23 passed**, covering gradient-value preservation, device round trips, partial conversion failures, and subsequent accumulation.
- Run on four GB200 GPUs using branch Python sources over existing PyTorch `2.13.0+cu130` binaries. A full current-main source build and broader CI validation remain outstanding.

Upstream regression command (at least two GPUs; four for HSDP):

```bash
python test/distributed/_composable/fsdp/test_fully_shard_grad_dtype.py
```

### CODE REVIEW

**Comment by xiuhu17** [comment]
@weifengpy would you like to take a look at this pr? It is a fix on wgrad precision issue. Thanks!

**Comment by xiuhu17** [comment]
Hi @weifengpy, hope you are doing well!

We are thinking it might be a design choice for PyTorch FSDP: A. whether have both wgrad dtype + reduction dtype or B. only wgrad dtype.

For A
          the pros: it is more backward compatible with current api. Use could choose to ignore how to set wgrad dtype by still using bf16. User could have lower precision on wgrad dtype, smaller memory usage while maintain relative higher precision on reduction part. 
          the cons: introduces confusion by having two semantic

For B
          the pros: semantic is more consistent.
          the cons: must choose the same wgrad dtype as reduction dtype. If user always choose fp32 on reduction part(I am guessing so), the memory usage is 2x.

**Comment by weifengpy** [comment]
> Hi @weifengpy, hope you are doing well!
> 
> 
> 
> We are thinking it might be a design choice for PyTorch FSDP: A. whether have both wgrad dtype + reduction dtype or B. only wgrad dtype.
> 
> 
> 
> For A
> 
>           the pros: it is more backward compatible with current api. Use could choose to ignore how to set wgrad dtype by still using bf16. User could have lower precision on wgrad dtype, smaller memory usage while maintain relative higher precision on reduction part. 
> 
>           the cons: introduces confusion by having two semantic
> 
> 
> 
> For B
> 
>           the pros: semantic is more consistent.
> 
>           the cons: must choose the same wgrad dtype as reduction dtype. If user always choose fp32 on reduction part(I am guessing so), the memory usage is 2x.
> 
> 

thanks for the design choices and drafted PR. we got the same problem from torchtitan recently.

I am repurposing mixed precision policy reduce_dtype to serve both reduction dtype and .grad_dtype for autograd: https://github.com/pytorch/pytorch/pull/194434 . It looks like option B

**Comment by xiuhu17** [comment]
> > Hi @weifengpy, hope you are doing well!
> > We are thinking it might be a design choice for PyTorch FSDP: A. whether have both wgrad dtype + reduction dtype or B. only wgrad dtype.
> > For A
> > ```
> >       the pros: it is more backward compatible with current api. Use could choose to ignore how to set wgrad dtype by still using bf16. User could have lower precision on wgrad dtype, smaller memory usage while maintain relative higher precision on reduction part. 
> > 
> >       the cons: introduces confusion by having two semantic
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > For B
> > ```
> >       the pros: semantic is more consistent.
> > 
> >       the cons: must choose the same wgrad dtype as reduction dtype. If user always choose fp32 on reduction part(I am guessing so), the memory usage is 2x.
> > ```
> 
> thanks for the design choices and drafted PR. we got the same problem from torchtitan recently.
> 
> I am repurposing mixed precision policy reduce_dtype to serve both reduction dtype and .grad_dtype for autograd: #194434 . It looks like option B

@weifengpy Thanks for this review. Should I close this pr in terms of the previous fix on https://github.com/pytorch/pytorch/pull/194434? Seems this design is well discussed within internal torch team?
